src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -65,14 +66,64 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
             combinedCancellationToken.ThrowIfCancellationRequested();
             _hostedServices = Services.GetRequiredService<IEnumerable<IHostedService>>();
 
-            foreach (IHostedService hostedService in _hostedServices)
+            List<Exception> exceptions = new List<Exception>();
+
+            if (_options.ServicesStartConcurrently)
             {
-                // Fire IHostedService.Start
-                await hostedService.StartAsync(combinedCancellationToken).ConfigureAwait(false);
+                Task tasks = Task.WhenAll(_hostedServices.Select(async service =>
+                {
+                    await service.StartAsync(combinedCancellationToken).ConfigureAwait(false);
+
+                    if (service is BackgroundService backgroundService)
+                    {
+                        _ = TryExecuteBackgroundServiceAsync(backgroundService);
+                    }
+                }));
 
-                if (hostedService is BackgroundService backgroundService)
+                try
                 {
-                    _ = TryExecuteBackgroundServiceAsync(backgroundService);
+                    await tasks.ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    exceptions.AddRange(tasks.Exception?.InnerExceptions?.ToArray() ?? new[] { ex });
```

Comment:
```suggestion
                    exceptions.AddRange(tasks.Exception?.InnerExceptions ?? new[] { ex });
```

There's no need to `ToArray()` here. Same comment for the stopping code.